studio/studio.py

```diff
@@ -237,6 +245,171 @@ async def export(q: Q):
+    return f'''
+```
+{output}
+```
+'''
+
+
+async def show_progress(q: Q, msg: str):
+    q.page['meta'].dialog.items = [
+        ui.progress(label=msg),
+        ui.expander(name='console_expander', label='Detail', items=[
+            ui.text(name='console_out', content=get_output(''), width='100%'),
+        ]),
+        ui.button(name='cancel_pip_task', label='Cancel')
+    ]
+    await q.page.save()
+
+
+async def update_progress(q: Q, value: int):
+    q.page['meta'].dialog.items[1].expander.items[0].text.content = get_output(value)
+    await q.page.save()
+
+
+async def show_finish_message(q: Q, type: Literal['error', 'success'], title: str, output: str):
+    q.page['meta'].dialog.items = [
+        ui.message_bar(type=type, text=title),
+        ui.expander(name='console_expander', label='Detail', items=[
+            ui.text(name='console_out', content=get_output(output), width='100%'),
+        ]),
+        ui.button(name='finish_message_dismiss', label='Go to package manager')
+    ]
+    await q.page.save()
+
+
+async def block_dialog(q: Q):
+    q.page['meta'].dialog.closable = False
+    q.page['meta'].dialog.blocking = True
+    await q.page.save()
+
+
+async def on_pip_finish(q: Q):
+    q.page['meta'].dialog.blocking = False
+    q.page['meta'].dialog.closable = True
+    q.page['meta'].dialog.items = get_package_dialog_items()
+    await q.page.save()
+
+
+async def pip(q: Q, command: str, flag: str or None, package_name: str, package_version: str, progress_message: str,
+              success_message: str, error_message: str, on_success: Callable = None, on_error: Callable = None):
+    try:
+        version = f'=={package_version}' if package_version else ''
+        args = [sys.executable, '-m', 'pip', command, flag, f'{package_name}{version}']
+        p = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output = ''
+        await block_dialog(q)
+        await show_progress(q, progress_message)
+        while True:
+            if p.poll() is not None:
+                # Process finished.
+                if p.returncode == 0:
+                    if on_success:
+                        await on_success(package_name)
+                    await show_finish_message(q, 'success', success_message, output)
+                else:
+                    if on_error:
+                        await on_error()
+                    output += f'\n{p.stderr.read().decode("utf-8")}'
+                    await show_finish_message(q, 'error', error_message, output)
+                break
+            else:
+                output += f'\n{p.stdout.readline().decode("utf-8")}'
+                await update_progress(q, output)
+    finally:
+        if p.poll() is None:
+            # Task canceled.
+            p.stdout.close()
+            p.stderr.close()
+            p.kill()
+            await on_pip_finish(q)
+
+
+def update_requirements(packages: dict, remove=False):
```

Comment:
Let's make this simpler:

* Open the file.
* Read and parse its packages.
* Add/remove the unnecessary package.
* Write (save) packages back to file.